File: exercises/obstacle_avoidance/web-template/lap.py
import rospy
from datetime import datetime

class Lap:
	def __init__(self, map_object):
		self.map = map_object
		
		self.target_start = "target01"
		self.target_end = "NaN"
		
		self.reset()
		
	# Function to check for threshold
	def check_threshold(self):
		target = self.map.getNextTarget()
		targetid = target.getId()
		
		if(targetid != self.target_end and targetid != self.target_start):
			if(self.buffer == True):
				self.start_time = datetime.now()
				self.buffer = False

			if(self.pause_condition == False):
				if(self.lap_time == 0):	
					self.lap_time = datetime.now() - self.start_time
				else:
					self.lap_time += datetime.now() - self.start_time
				
				self.start_time = datetime.now()
	
		return self.lap_time
			
			
	# Function to return lap time
	def return_lap_time(self):
		return self.lap_time

	# Function to reset
	def reset(self):
		# Reset the initial variables
		self.start_time = 0
		self.lap_time = 0

		self.buffer = True
		self.pause_condition = False

	# Function to pause
	def pause(self):
		self.pause_condition = True

	# Function to unpause
	def unpause(self):
		# To enable unpause button to be used again and again
		if(self.pause_condition == True):
			# Overall subtract from lap_time, the time we have been paused
			self.start_time = datetime.now()

		self.pause_condition = False
		   
		   
		   
		   
		   
		   
			
		

File: exercises/obstacle_avoidance/cpp_lib/src/Lap.cpp
#include "Lap.hpp"

Lap::Lap(std::shared_ptr<Map> map_object) 
    : map_(map_object), target_start_("target01"), target_end_("NaN") {
    reset();
}

double Lap::check_threshold() {
    auto target = map_->getNextTarget();
    if (!target) return lap_time_.count();

    std::string targetid = target->getId();

    if (targetid != target_end_ && targetid != target_start_) {
        if (buffer_) {
            start_time_ = std::chrono::system_clock::now();
            buffer_ = false;
        }

        if (!pause_condition_) {
            auto now = std::chrono::system_clock::now();
            if (lap_time_.count() == 0.0) {
                lap_time_ = now - start_time_;
            } else {
                lap_time_ += now - start_time_;
            }
            start_time_ = now;
        }
    }

    return lap_time_.count();
}

double Lap::return_lap_time() const {
    return lap_time_.count();
}

void Lap::reset() {
    lap_time_ = std::chrono::duration<double>::zero();
    buffer_ = true;
    pause_condition_ = false;
}

void Lap::pause() {
    pause_condition_ = true;
}

void Lap::unpause() {
    if (pause_condition_) {
        start_time_ = std::chrono::system_clock::now();
    }
    pause_condition_ = false;
}